src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java

```diff
@@ -0,0 +1,142 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+class ClientConnectionManager {
+  private final ClientBuilder builder;
+  private final AtomicReference<ManagedChannel> channelRef;
+  private final AtomicReference<Optional<String>> tokenRef;
+  private final ExecutorService executorService;
+
+  ClientConnectionManager(ClientBuilder builder) {
+    this(builder, Executors.newCachedThreadPool(), null);
+  }
+
+  ClientConnectionManager(
+      ClientBuilder builder, ExecutorService executorService) {
+
+    this(builder, executorService, null);
+  }
+
+  ClientConnectionManager(
+      ClientBuilder builder, ManagedChannel channel) {
+
+    this(builder, Executors.newCachedThreadPool(), channel);
+  }
+
+  ClientConnectionManager(
+      ClientBuilder builder, ExecutorService executorService, ManagedChannel channel) {
+
+    this.builder = builder;
+    this.channelRef = new AtomicReference<>(channel);
+    this.tokenRef = new AtomicReference<>();
+    this.executorService = executorService;
+  }
+
+  ManagedChannel getChannel() {
+    ManagedChannel mc = channelRef.get();
+    if (mc == null) {
+      synchronized (channelRef) {
+        mc = channelRef.get();
+        if (mc == null) {
+          NameResolver.Factory resolverFactory = builder.getNameResolverFactory();
+          if (resolverFactory == null) {
+            resolverFactory = ClientUtil.simpleNameResolveFactory(builder.getEndpoints());
+          }
+
+          ManagedChannelBuilder<?> channelBuilder = builder.getChannelBuilder();
+          if (channelBuilder == null) {
+            channelBuilder = defaultChannelBuilder(resolverFactory);
+          }
+
+          mc = channelBuilder.build();
+          channelRef.lazySet(mc);
+        }
+      }
+    }
+
+    return mc;
+  }
+
+  Optional<String> getToken() {
+    Optional<String> tk = tokenRef.get();
+    if (tk == null) {
+      synchronized (tokenRef) {
+        tk = tokenRef.get();
+        if (tk == null) {
+          tk = ClientUtil.generateToken(getChannel(), builder.getUser(), builder.getPassword());
+          tokenRef.lazySet(tk);
+        }
+      }
+    }
+
+    return tk;
+  }
+
+  ExecutorService getExecutorService() {
+    return executorService;
+  }
+
+  /**
+   * create and add token to channel's head.
+   *
+   * @param supplier the stub supplier
+   * @param <T> the type of stub
+   * @return the attached stub
+   */
+  <T extends AbstractStub<T>> T newStub(Function<ManagedChannel, T> supplier) {
+    return ClientUtil.configureStub(supplier.apply(getChannel()), getToken());
+  }
+
+  synchronized void close() {
+    ManagedChannel channel = channelRef.get();
+    if (channel != null) {
+      channel.shutdown();
+    }
+
+    executorService.shutdownNow();
+  }
+
+  <T extends AbstractStub<T>, R> CompletableFuture<R> withNewChannel(
+      String endpoint,
+      Function<ManagedChannel, T> stubCustomizer,
+      BiFunction<ManagedChannel, T, CompletableFuture<R>> function) {
```

Comment:
` BiFunction<ManagedChannel, T, CompletableFuture<R>> function) ` - > 
` Function< T, CompletableFuture<R>> function) `
I didn't see any usage for `ManagedChannel` arg in `Maintenance.status()` and  `Maintenance.defragmentMember() `. So we probably don't need it.